aip/0162.md

```diff
@@ -0,0 +1,312 @@
+when to commit a new revision, such as:
+
+- Commit a new revision any time that there is a change
+- Commit a new revision when something important happens
+- Commit a new revision when the user specifically asks
+
+APIs **may** use any of these strategies. APIs that want to commit a revision
+on user request **should** handle this with a `Commit` custom method:
+
+```proto
+rpc CommitBook(CommitBookRequest) returns (Book) {
+  option (google.api.http) = {
+    post: "/v1/{name=publishers/*/books/*}:commit"
+    body: "*"
+  }
+}
+```
+
+- The method **must** use the `POST` HTTP verb.
+- The method **should** return the resource, and the resource name **must**
+  include the revision ID.
+
+The request message **must** include the `name` field, and **may** include any
+additional metadata about the revision (such as a description), which
+**should** then be included in revision objects.
+
+### Rollback
+
+A common use case for a resource with a revision history is the ability to roll
+back to a given revision. APIs **should** handle this with a `Rollback` custom
+method:
+
+```proto
+rpc RollbackBook(RollbackBookRequest) returns (Book) {
+  option (google.api.http) = {
+    post: "/v1/{name=publishers/*/books/*}:rollback"
+    body: "*"
+  }
+}
+```
+
+- The method **must** use the `POST` HTTP verb.
+- The method **should** return the resource, and the resource name **must**
+  include the revision ID.
+
+```proto
+message RollbackBookRequest {
+  // The book being rolled back.
+  string name = 1;
+
+  // The revision ID to roll back to.
+  // It must be a revision of the same book.
+  //
+  //   Example: c7cfa2a8
+  string revision_id = 2;
+}
+```
+
+- The request message **must** include a `revision_id` field.
+  - The API **must** fail the request with `NOT_FOUND` if the revision does not
+    exist on that resource.
+
+**Note:** When rolling back, the API should return a _new_ revision of the
+resource with a _new_ revision ID, rather than reusing the original ID. This
+avoids problems with representing the same revision being active for multiple
+ranges of time.
+
+### Deleting revisions
+
+Revisions are sometimes expensive to store, and there are valid use cases to
+want to remove one or more revisions from a resource's revision history.
+
+APIs **may** define a method to delete revisions, with a structure similar to
+`Delete` ([AIP-135][]) methods:
+
+```proto
+rpc DeleteBookRevision(DeleteBookRevisionRequest)
+    returns (google.protobuf.Empty) {
+  option (google.api.http) = {
+    delete: "/v1/{name=publishers/*/books/*}:deleteRevision"
+  }
+}
+```
+
+```proto
+message DeleteBookRevisionRequest {
+  // The name of the book revision to be deleted, with a revision ID explicitly
+  // included.
+  //
+  // Example: publishers/123/books/les-miserables@c7cfa2a8
+  string name = 1;
+}
+```
+
+- The explicit resource ID **must** be required (the method **must** fail with
```

Comment:
Worth talking about a BatchDeleteBookRevisions custom method? There, it would make sense to specify a single name, but repeated revisions. If that seems reasonable, should it affect this case as well?

Reply:
I am inclined to wait until someone actually need this. Thoughts?